File: src/1382-Balance-a-Binary-Search-Tree/1382.go
var res []int

func balanceBST(root *TreeNode) *TreeNode {
    res = []int{}
    inOrder(root)
    return sortedToBST(0, len(res) - 1)
}

func inOrder(root *TreeNode) {
    if root != nil {
        inOrder(root.Left)
        res = append(res, root.Val)
        inOrder(root.Right)
    }
}

func sortedToBST(l, r int) *TreeNode {
    if l > r {
        return nil
    }
    
    mid := (l + r) >> 1
    root := new(TreeNode)
    root.Val = res[mid]
    root.Left = sortedToBST(l, mid - 1)
    root.Right = sortedToBST(mid + 1, r)
    return root
}

File: src/1382-Balance-a-Binary-Search-Tree/1382.js
var balanceBST = function(root) {
    let res = [];
    let inOrder = function(root) {
        if (root != null) {
            inOrder(root.left);
            res.push(root.val);
            inOrder(root.right);
        }
    }
    
    inOrder(root);
    
    let sortedToBST = function(l, r) {
        if (l > r) return  null;
        let mid = (l + r) >> 1;
        let root = new TreeNode(res[mid]);
        root.left = sortedToBST(l, mid - 1);
        root.right = sortedToBST(mid + 1, r);
        return root;
    }
    return sortedToBST(0, res.length - 1);  
};